server: TestAddNewStoresToExistingNodes failed

server.TestAddNewStoresToExistingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10886615?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10886615?buildTab=artifacts#/) on master @ [665084ed379a25fcde92ee8b4c9dc48e192876e5](https://github.com/cockroachdb/cockroach/commits/665084ed379a25fcde92ee8b4c9dc48e192876e5):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 131489 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestAddNewStoresToExistingNodes
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/dbd7076bfe20e3183065a44fa4f7e406/logTestAddNewStoresToExistingNodes363450179
    test_log_scope.go:81: use -show-logs to present logs inline
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv @cockroachdb/storage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAddNewStoresToExistingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestAddNewStoresToExistingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10949923?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10949923?buildTab=artifacts#/) on master @ [b3ae46d39098088ca33844b3cb87d18c60cde0d5](https://github.com/cockroachdb/cockroach/commits/b3ae46d39098088ca33844b3cb87d18c60cde0d5):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 117612 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestAddNewStoresToExistingNodes
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/dbd7076bfe20e3183065a44fa4f7e406/logTestAddNewStoresToExistingNodes3219261074
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:104: 
        Test server was configured to route SQL queries to a secondary tenant (virtual cluster).
        If you are only seeing a test failure when this message appears, there may be a problem
        specific to cluster virtualization or multi-tenancy.
        
        To investigate, consider using "COCKROACH_TEST_TENANT=true" to force-enable just
        the secondary tenant in all runs (or, alternatively, "false" to force-disable), or use
        "COCKROACH_INTERNAL_DISABLE_METAMORPHIC_TESTING=false" to disable all random test variables altogether.
*
* ERROR: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
*
* ERROR: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
*
* ERROR: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
    test_server_shim.go:104: 
        Test server was configured to route SQL queries to a secondary tenant (virtual cluster).
        If you are only seeing a test failure when this message appears, there may be a problem
        specific to cluster virtualization or multi-tenancy.
        
        To investigate, consider using "COCKROACH_TEST_TENANT=true" to force-enable just
        the secondary tenant in all runs (or, alternatively, "false" to force-disable), or use
        "COCKROACH_INTERNAL_DISABLE_METAMORPHIC_TESTING=false" to disable all random test variables altogether.
    testcluster.go:415: 
        Test server was configured to route SQL queries to a secondary tenant (virtual cluster).
        If you are only seeing a test failure when this message appears, there may be a problem
        specific to cluster virtualization or multi-tenancy.
        
        To investigate, consider using "COCKROACH_TEST_TENANT=true" to force-enable just
        the secondary tenant in all runs (or, alternatively, "false" to force-disable), or use
        "COCKROACH_INTERNAL_DISABLE_METAMORPHIC_TESTING=false" to disable all random test variables altogether.
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAddNewStoresToExistingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/liveness/BUILD.bazel
+++ b/pkg/kv/kvserver/liveness/BUILD.bazel
@@ -24,6 +24,7 @@ go_library(
         "//pkg/settings",
         "//pkg/settings/cluster",
         "//pkg/storage",
+        "//pkg/util/buildutil",
         "//pkg/util/grpcutil",
         "//pkg/util/hlc",
         "//pkg/util/log",

--- a/pkg/kv/kvserver/liveness/liveness.go
+++ b/pkg/kv/kvserver/liveness/liveness.go
@@ -28,6 +28,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/settings"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	diskStorage "github.com/cockroachdb/cockroach/pkg/storage"
+	"github.com/cockroachdb/cockroach/pkg/util/buildutil"
 	"github.com/cockroachdb/cockroach/pkg/util/grpcutil"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -732,6 +733,16 @@ var errNodeAlreadyLive = errors.New("node already live")
 // by the Start loop.
 // TODO(bdarnell): Should we just remove this synchronous heartbeat completely?
 func (nl *NodeLiveness) Heartbeat(ctx context.Context, liveness livenesspb.Liveness) error {
+	if buildutil.CrdbTestBuild && !nl.started.Get() {
+		// This check was added as part of resolving #106706. We were previously
+		// accidentally relying on synchronous heartbeats to paper over problems,
+		// which only worked most of the time but could lead to hangs.
+		// In our test builds, we only allow heartbeats of any kind once the
+		// liveness loop has started.
+		//
+		// See: https://github.com/cockroachdb/cockroach/issues/106706#issuecomment-1640254715
+		return errors.New("liveness heartbeat not started yet")
+	}
 	return nl.heartbeatInternal(ctx, liveness, false /* increment epoch */)
 }
 

--- a/pkg/server/server.go
+++ b/pkg/server/server.go
@@ -1910,13 +1910,17 @@ func (s *Server) PreStart(ctx context.Context) error {
 	// initialized.
 	s.grpc.setMode(modeOperational)
 
+	s.nodeLiveness.Start(workersCtx)
+
 	// We'll block here until all stores are fully initialized. We do this here
-	// for two reasons:
+	// for several reasons:
 	// - some of the components below depend on all stores being fully
 	//   initialized (like the debug server registration for e.g.)
 	// - we'll need to do it after having opened up the RPC floodgates (due to
 	//   the hazard described in Node.start, around initializing additional
 	//   stores)
+	// - we'll need to do it after starting node liveness, see:
+	//   https://github.com/cockroachdb/cockroach/issues/106706#issuecomment-1640254715
 	s.node.waitForAdditionalStoreInit()
 
 	additionalStoreIDs, err := state.additionalStoreIDs()
@@ -1963,11 +1967,6 @@ func (s *Server) PreStart(ctx context.Context) error {
 	log.Ops.Infof(ctx, "advertising CockroachDB node at %s", log.SafeManaged(s.cfg.AdvertiseAddr))
 
 	log.Event(ctx, "accepting connections")
-
-	// Begin the node liveness heartbeat. Add a callback which records the local
-	// store "last up" timestamp for every store whenever the liveness record is
-	// updated.
-	s.nodeLiveness.Start(workersCtx)
 
 	// Begin recording status summaries.
 	if err := s.node.startWriteNodeStatus(base.DefaultMetricsSampleInterval); err != nil {

